### PR TITLE
Pass call type in from/toGraphQL

### PR DESCRIPTION
**Pull request by papigers**
Currently types are ignored and not getting passed in `callsFromGraphQL`
or `callsToGraphQL`.
This results in dangerous output from `toGraphQL`, in which all calls have
type null. This can cause problems when using non-scalar types (like
enum).
For more details see #1256.

### CODE REVIEW

**Comment by facebook-github-bot-1** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-8** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: src/query/callsFromGraphQL.js [hunk 1]
```diff
@@ -42,7 +43,11 @@ function callsFromGraphQL(
   const orderedCalls = [];
   for (let ii = 0; ii < callsOrDirectives.length; ii++) {
     const callOrDirective = callsOrDirectives[ii];
-    let {value} = callOrDirective;
+    let {value, metadata} = callOrDirective;
+    let type = undefined;
```
Comment (josephsavona) [inline src/query/callsFromGraphQL.js]: `let type;`

File: src/query/callsFromGraphQL.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ const invariant = require('invariant');
 type CallOrDirective = {
   name: string,
   value: ?ConcreteValue,
+  type?: ?string,
```
Comment (josephsavona) [inline src/query/callsFromGraphQL.js]: Remove this? The code never accesses `type` on the `callsOrDirectives` array.

**Comment by josephsavona** [comment]
@papigers Thanks for the PR. Did you run tests locally? The build is not failing, the tests are ;-)

File: src/query/callsFromGraphQL.js (hunk 2 above)
Comment (papigers) [inline src/query/callsFromGraphQL.js]: True, my bad

File: src/query/callsFromGraphQL.js (hunk 1 above)
Comment (papigers) [inline src/query/callsFromGraphQL.js]: :+1:

**Comment by papigers** [comment]
@josephsavona Will do, working on that right now, sorry for this noobish PR.

Next time will be better, I hope ;)

**Comment by papigers** [comment]
Sorry for the uninformatives commits

File: src/tools/RelayInternalTypes.js [hunk 3]
```diff
@@ -23,7 +23,7 @@ import type RelayQuery from 'RelayQuery';
 
 export type Call = {
   name: string,
-  type?: string,
+  type?: ?string,
```
Comment (steveluscher) [inline src/tools/RelayInternalTypes.js]: I haven't thought about this as deeply as you two have, but might we instead want `type` to either be a `string` or not supplied?

```
const orderedCall = {
  name: callOrDirective.name, 
  value,
};
if (metadata && metadata.type) {
  orderedCall.type = metadata.type;
}
orderedCalls.push(orderedCall);
```

That would cut down on the noise in the tests, among other things. If it's important to have `type: null` then, carry on!

Comment (josephsavona) [inline src/tools/RelayInternalTypes.js]: Yeah, this already is marked as an optional field.

@papigers I apologize for the back and forth on this. Since this is already an optional field, how about reverting the addition of all the `type: null`s to focus this PR on just the logic & results that actually change?

File: src/query/callsFromGraphQL.js [hunk 4]
```diff
@@ -23,9 +23,14 @@ import type {Variables} from 'RelayTypes';
 
 const invariant = require('invariant');
 
+type Metadata = {
+  type: ?string,
```
Comment (josephsavona) [inline src/query/callsFromGraphQL.js]: let's make this `type?: ?string`

**Comment by josephsavona** [comment]
@papigers Thanks again for this PR. I commented to request some changes. Also, scanning through i don't see any tests that actually end up with `type` having a non-null value, which means it doesn't look like we're really testing the change itself. 

I apologize if I missed something, but if not can you add a test that has a non-null type? This would be an enum or object argument.

**Comment by psalz** [comment]
I just ran into the same issue - any updates on this?

**Comment by josephsavona** [comment]
@papigers are you able to update the PR?

**Comment by papigers** [comment]
Not any time soon , I'm afraid. Too busy..

**Comment by josephsavona** [comment]
closing in favor of #1336 - thanks @papigers!
